tensorflow_probability/python/distributions/gev.py

```diff
@@ -0,0 +1,265 @@
+          loc, name='loc', dtype=dtype)
+      scale = tensor_util.convert_nonref_to_tensor(
+          scale, name='scale', dtype=dtype)
+      concentration = tensor_util.convert_nonref_to_tensor(
+          concentration, name='concentration', dtype=dtype)
+      dtype_util.assert_same_float_dtype([loc, scale, concentration])
+      # Positive scale is asserted by the incorporated GEV bijector.
+      self._gev_bijector = gev_cdf_bijector.GeneralizedExtremeValueCDF(
+          loc=loc, scale=scale, concentration=concentration,
+          validate_args=validate_args)
+
+      # Because the uniform sampler generates samples in `[0, 1)` this would
+      # cause samples to lie in `(inf, -inf]` instead of `(inf, -inf)`. To fix
+      # this, we use `np.finfo(dtype_util.as_numpy_dtype(self.dtype).tiny`
+      # because it is the smallest, positive, 'normal' number.
+      batch_shape = distribution_util.get_broadcast_shape(loc, scale,
+                                                          concentration)
+      super(GeneralizedExtremeValue, self).__init__(
+          # TODO(b/137665504): Use batch-adding meta-distribution to set the
+          # batch shape instead of tf.ones.
+          distribution=uniform.Uniform(
+              low=np.finfo(dtype_util.as_numpy_dtype(dtype)).tiny,
+              high=tf.ones(batch_shape, dtype=dtype),
+              allow_nan_stats=allow_nan_stats),
+          # The GEV bijector encodes the CDF function as the forward,
+          # and hence needs to be inverted.
+          bijector=invert_bijector.Invert(
+              self._gev_bijector, validate_args=validate_args),
+          parameters=parameters,
+          name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return dict(
+        zip(('loc', 'scale', 'concentration'),
+            ([tf.convert_to_tensor(sample_shape, dtype=tf.int32)] * 3)))
+
+  @classmethod
+  def _params_event_ndims(cls):
+    return dict(loc=0, scale=0, concentration=0)
+
+  @property
+  def loc(self):
+    """Distribution parameter for the location."""
+    return self._gev_bijector.loc
+
+  @property
+  def scale(self):
+    """Distribution parameter for scale."""
+    return self._gev_bijector.scale
+
+  @property
+  def concentration(self):
+    """Distribution parameter for shape."""
+    return self._gev_bijector.concentration
+
+  def _entropy(self):
+    # Use broadcasting rules to calculate the full broadcast sigma.
+    scale = self.scale * tf.ones_like(self.loc)
+    return 1. + tf.math.log(scale) + np.euler_gamma*(1. + self.concentration)
+
+  def _log_prob(self, x):
+    with tf.control_dependencies(self._gev_bijector._maybe_assert_valid_x(x)):
+      scale = tf.convert_to_tensor(self.scale)
+      z = (x - self.loc) / scale
+
+      conc = tf.convert_to_tensor(self.concentration)
+      equal_zero = tf.equal(conc, 0.)
+      log_t = tf.where(equal_zero, -z,
+                       -tf.math.log1p(z * conc) / conc)
```

Comment:
This needs to be written somewhat differently, because as is it'll produce NaN gradients when conc is 0, see https://github.com/tensorflow/probability/blob/master/discussion/where-nan.pdf for details. In this case, what you can do it something like:

```python
safe_conc = tf.where(equal_zero, 1., conc)
log_t = tf.where(equal_zero, -z,
        -tf.math.log1p(z * safe_conc) / safe_conc)
```